src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,493 @@
+use Symfony\Component\Cache\Adapter\TagAwareAdapterInterface;
+use Symfony\Component\Cache\PruneableInterface;
+use Symfony\Component\HttpFoundation\HeaderBag;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpCache\StoreInterface;
+use Symfony\Component\Lock\Exception\LockReleasingException;
+use Symfony\Component\Lock\Factory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Component\Lock\Store\FlockStore;
+use Symfony\Component\Lock\Store\SemaphoreStore;
+
+/**
+ * Implements a storage for Symfony's HttpCache that supports tagging.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class TaggableStore implements StoreInterface
+{
+    const NON_VARYING_KEY = 'non-varying';
+    const COUNTER_KEY = 'write-operations-counter';
+
+    /**
+     * @var string
+     */
+    private $purgeTagsHeader;
+
+    /**
+     * @var int
+     */
+    private $pruneThreshold;
+
+    /**
+     * @var TagAwareAdapterInterface
+     */
+    private $cache;
+
+    /**
+     * @var Factory
+     */
+    private $lockFactory;
+
+    /**
+     * @var LockInterface[]
+     */
+    private $locks = [];
+
+    /**
+     * Constructor.
+     *
+     * @param string $cacheDir        The cache directory
+     * @param int    $pruneThreshold  The number of write operations until orphan data gets pruned (default: 500)
+     * @param string $purgeTagsHeader The tags header name
+     */
+    public function __construct($cacheDir, $pruneThreshold = 500, $purgeTagsHeader = PurgeTagsListener::DEFAULT_PURGE_TAGS_HEADER)
+    {
+        if (!class_exists(Factory::class)) {
+            throw new \RuntimeException('You need at least Symfony 3.4 to use the TaggableStore.');
+        }
+
+        $this->purgeTagsHeader = $purgeTagsHeader;
+        $this->pruneThreshold = $pruneThreshold;
+
+        // Default lifetime does not need to be configurable because we're working on a reverse proxy cache here
+        // which means the cache entries (= responses) MUST have cache related headers so we have individual
+        // expiry dates for each entry.
+        $this->setCache(new TagAwareAdapter(new FilesystemAdapter('fos-http-cache', 0, $cacheDir)));
+        $this->setLockFactory(new Factory($this->getBestLocalLockStore($cacheDir)));
+    }
+
+    /**
+     * Use this method if you want to use a different cache implementation than
+     * the one created in the constructor. Note that there are very good reasons
+     * that the local adapters are used by default and you do not inject them in
+     * the constructor but via setter injection instead. This is to protect you
+     * as a developer! Only use this method if you're really sure your cache
+     * implementation meets the needs of Symfony's HttpCache.
+     *
+     * @param TagAwareAdapterInterface $cache
+     *
+     * @return TaggableStore
+     */
+    public function setCache(TagAwareAdapterInterface $cache)
```

Comment:
lets move this into the options. with the current implementation, the file system adapter is created and then replaced, which is just odd. i agree that we put a warning there - we should mention the "very good reasons" however, so that people can make an informed decision ;-)